profile/profile.go

```diff
@@ -1,241 +1,311 @@
+func (p *Profile) EthereumAddress() (common.Address, error) {
 
-	// id public key
-	idPubKeyStr, err := km.IdentityPublicKey()
+	// validate profile signature
+	valid, err := p.SignaturesValid()
 	if err != nil {
-		return Profile{}, err
+		return common.Address{}, err
 	}
-	idPubKey, err := hex.DecodeString(idPubKeyStr)
-	if err != nil {
-		return Profile{}, err
+	if !valid {
+		return common.Address{}, errors.New("invalid profile signature")
 	}
 
-	// ethereum public key
-	ethPubKeyStr, err := km.GetEthereumPublicKey()
-	if err != nil {
-		return Profile{}, err
-	}
-	ethPubKey, err := hex.DecodeString(ethPubKeyStr)
+	// raw ethereum key
+	ethPubKey, err := p.EthereumPublicKey()
 	if err != nil {
-		return Profile{}, err
+		return common.Address{}, err
 	}
 
-	// now
-	now := time.Now().UTC()
-
-	// chat id public key
-	pair, err := km.ChatIdKeyPair()
+	// decompressed ethereum key
+	decPubKey, err := crypto.DecompressPubkey(ethPubKey)
 	if err != nil {
-		return Profile{}, err
+		return common.Address{}, err
 	}
-	chatPub := pair.PublicKey
 
-	// concat profile information
-	dataToSign := []byte(name)
-	dataToSign = append(dataToSign, []byte(location)...)
-	dataToSign = append(dataToSign, []byte(image)...)
-	dataToSign = append(dataToSign, []byte(now.String())...)
-	dataToSign = append(dataToSign, idPubKey...)
-	dataToSign = append(dataToSign, ethPubKey...)
-	dataToSign = append(dataToSign, chatPub[:]...)
-	version := make([]byte, 2)
-	binary.LittleEndian.PutUint16(version, profileVersion)
-	dataToSign = append(dataToSign, version...)
+	return crypto.PubkeyToAddress(*decPubKey), nil
+}
 
-	// hash of profile data
-	dataHash := sha256.Sum256(dataToSign)
+// sign the metadata with identity and ethereum key
+func SignProfile(name, location, image string, km km.KeyManager) (*Profile, error) {
 
-	// hash the profile data
+	// chat id key pair
+	chatIDKeys, err := km.ChatIdKeyPair()
 	if err != nil {
-		return Profile{}, err
+		return nil, err
 	}
 
-	// sign hash of data with id key
-	idKeySignature, err := km.IdentitySign(dataHash[:])
+	// identity pub key
+	idPubKeyStr, err := km.IdentityPublicKey()
+	if err != nil {
+		return nil, err
+	}
+	idPubKeyRaw, err := hex.DecodeString(idPubKeyStr)
 	if err != nil {
-		return Profile{}, err
+		return nil, err
 	}
 
-	// sign hash of data with eth key
-	ethKeySignature, err := km.EthereumSign(dataHash)
+	// ethereum pub key
+	ethPubKeyStr, err := km.GetEthereumPublicKey()
 	if err != nil {
-		return Profile{}, err
+		return nil, err
+	}
+	ethPubKeyRaw, err := hex.DecodeString(ethPubKeyStr)
+	if err != nil {
+		return nil, err
 	}
 
-	// profile
-	prof := Profile{
+	p := Profile{
 		Information: Information{
 			Name:           name,
 			Location:       location,
 			Image:          image,
-			IdentityPubKey: idPubKeyStr,
-			EthereumPubKey: ethPubKeyStr,
-			Timestamp:      now,
-			ChatIDKey:      chatPub,
+			ChatIDKey:      chatIDKeys.PublicKey,
+			Timestamp:      time.Now(),
```

Comment:
Might be preferable to use time.Now().UnixNano() instead of time.Now() as UnixNano leaves almost no room for error if we ever need to covert to different timezones or time formats.

Reply:
I switched to unix timestamp. Please review :)